LoggingService: Improve tests by minimising implementation details

AudreyFelicio opened the pull request:
# Summary 

Fixes #645 

## Description 

We should minimise usage of private methods and implementation details in the tests.
This will make it easier to refactor LoggingService in future, and the tests become more representative of how the service is used.
Currently, the tests use the internal getTrimmedLogCache method directly, and they rely on implementation details
(e.g. how to determine the number of sessions logged, by parsing the whole logs' string)
We could restructure the tests as follows:
- Mock localStorage using Jasmine.
- As far as possible, only use the public API of LoggingService (i.e. `.startSession()` and `.reset()` to simulate new sessions, `.info()` / `.error()` to add logs).
- Match the stored log against an expected log (without any parsing logic in the test code, as done currently).

Source: as discussed [here](https://github.com/CATcher-org/CATcher/pull/616#discussion_r592088052)

## Changes Made 

- Removed `getFilteredLogCount` which heavily relies on inner implementations.
- Rewrite and add more tests that compare actual and expected logs without parsing code.
- Make internal methods `private`.

## Proposed Commit Message 
```
Improve Logging Service Testing

Currently, a lot of implementation details are used
in the logging service tests.

Refactor the tests, get rid of private methods in tests,
and use public methods for testing by comparing actual
and expected logs.
```

AudreyFelicio commented on src/app/core/services/logging.service.ts:
```diff
@@ -53,7 +53,7 @@ export class LoggingService {
    */
-  getTrimmedLogCache(currentLog: string, sessionCount: number): string {
+  private getTrimmedLogCache(currentLog: string, sessionCount: number): string {
     const sessionLogSeparator: string = '\n'.repeat(2); // More new-lines added for clarity.
```
Just a suggestion:

Should we make this `sessionLogSeparator` a `public readonly` class field instead of local variable inside `getTrimmedLogCache`?

By making it a class field, we can then use it in the testing file as
`const sessionSeparator = loggingService.SESSION_LOG_SEPARATOR;`
In my opinion, this will make the tests more extensible and general as changes to the session log separator in the implementation will no longer break the tests.

dingyuchen commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
+    });
+
+    it('should successfully reinitialize logging session when limit reached', () => {
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT; i += 1) {
+        loggingService.startSession();
+        loggingService.reset();
+      }
```
```suggestion
      Array(loggingService.LOG_COUNT_LIMIT).fill().forEach(() => {
        loggingService.startSession();
        loggingService.reset();
      });
```

dingyuchen replied:
I think it is not really necessary to follow my suggestion but the declarative style might be a little better to read.

dingyuchen commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
+      }
+      loggingService.startSession();
+      const actualLog = loggingService.getCachedLog();
+      let expectedLog = '';
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT; i += 1) {
+        if (i === loggingService.LOG_COUNT_LIMIT - 1) {
+          expectedLog += headerLog;
+        } else {
+          expectedLog += `${headerLog}${sessionSeparator}`;
+        }
+      }
```
```suggestion
      const expectedLog = [...Array(loggingService.LOG_COUNT_LIMIT)].map( _ => headerLog).join(sessionSeparator);
```
A declarative style would definitely be better here

AudreyFelicio replied:
It definitely looks more concise and clear, thanks for the suggestion!

dingyuchen commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
-      // Number of Old Logs must be reduced by 1 (To make way for the new session log)
-      expect(getFilteredLogCount(repeatOldLogStartHeader(loggingService.LOG_COUNT_LIMIT), oldLogFilter)).toEqual(
-        loggingService.LOG_COUNT_LIMIT - 1
-      );
+    it('should successfully add error logs', () => {
+      testAddLog(errorLogMessage);
     });
 
-    it('should return trimmed and updated log if log in cache exceeds max number of sessions', () => {
-      const exceededSessionCount = loggingService.LOG_COUNT_LIMIT + 10; // Arbitrary Exceed Count
+    it('should successfully add warn logs', () => {
+      testAddLog(warnLogMessage);
+    });
 
-      // Number of logs must be at max number
-      expect(getFilteredLogCount(repeatOldLogStartHeader(exceededSessionCount), logHeaderFilter)).toEqual(loggingService.LOG_COUNT_LIMIT);
+    it('should successfully add debug logs', () => {
+      testAddLog(debugLogMessage);
+    });
+  });
```
Since `testAddLog` runs `loggingService.info` for all cases, perhaps it would be better to pass the logging function (eg `loggingService.debug/error/warn`) as a second argument.
Even so, all paths call `updateLogs`, which does not improve code coverage significantly. Inputs are also of the same equivalence class. Perhaps we can remove these tests altogether. What do the rest think? @CATcher-org/2021-devs

AudreyFelicio replied:
@dingyuchen 
Should we test the `updateLogs` function instead?
I've made the `updateLogs` method `private` since based on my understanding, we're supposed to call these 4 methods to log stuffs.

dingyuchen replied:
I think making `updateLogs` private is the right call, but I would argue that testing _one_ of `loggingService.(info|debug|error|warn)` is sufficient
ie just `testAddLog` once

dingyuchen commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
+      }
+      loggingService.startSession();
+      const actualLog = loggingService.getCachedLog();
+      let expectedLog = '';
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT; i += 1) {
+        if (i === loggingService.LOG_COUNT_LIMIT - 1) {
+          expectedLog += headerLog;
+        } else {
+          expectedLog += `${headerLog}\n${infoLogMessage}${sessionSeparator}`;
+        }
+      }
```
Similar to above

dingyuchen reviewed (commented):
Thanks for working on the Logging Service! Unfortunately I do not remember all the implementation details as much as I used to so feel free to comment if I got anything wrong 😄

AudreyFelicio commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
 describe('LoggingService', () => {
-  beforeAll(() => {
+  beforeEach(() => {
+    // Initialize mock local storage
+    let mockLocalStorage = {};
+    const mockLocalStorageFunctions = {
+      getItem: (key: string): string => {
+        return key in mockLocalStorage ? mockLocalStorage[key] : null;
+      },
+      setItem: (key: string, value: string) => {
+        mockLocalStorage[key] = value;
+      },
+      removeItem: (key: string) => {
+        delete mockLocalStorage[key];
+      },
+      clear: () => {
+        mockLocalStorage = {};
+      }
+    };
```
@dingyuchen
I think this is a good idea. Thanks for the suggestion!

AudreyFelicio commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,160 @@
-      expect(getFilteredLogCount('', logHeaderFilter)).toEqual(1);
-      expect(getFilteredLogCount('gibberish', logHeaderFilter)).toEqual(1);
+  afterEach(() => {
+    loggingService.reset();
+    localStorage.clear();
+    jasmine.clock().uninstall();
```
Turns out that we need to cleanup (uninstall jasmine clock) in `afterAll`, otherwise some other tests in other files would break due to timeout errors (due to the hardcoded clock)

AudreyFelicio commented:
@dingyuchen 
Thank you for the reviews and apologies for the very late update.
I've implemented your suggestions and also moved the `mockLocalStorage` into a separate `class` under `helper` folder, so that it's usable in other tests as well.

Can you help to re-review and check if all the changes are okay now?
Thanks!

dingyuchen commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,128 @@
-      expect(getFilteredLogCount(repeatOldLogStartHeader(exceededSessionCount), oldLogFilter)).toEqual(loggingService.LOG_COUNT_LIMIT - 1);
+  describe('updating and trimming logs from sessions', () => {
+    it('should trim oldest log if number of sessions exceed session limit', () => {
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT + 1; i += 1) {
```
you may use a forEach loop here

AudreyFelicio replied:
Oh yeah I missed this, thanks for the suggestion!

dingyuchen reviewed (approved):
Thanks for your hard work! I only have a small nit, other than that LGTM

anubh-v reviewed (commented):
Looks good to me, just 2 comments
Thanks @dingyuchen for the thorough review as well :+1:

AudreyFelicio commented on tests/services/logging.service.spec.ts:
```diff
@@ -1,64 +1,130 @@
+const mockDate = new Date(2021, 6, 27);
+const infoLogMessage = 'Info log message';
+
+const testAddLog = (message: string) => {
```
@anubh-v 
It's because before the review changes, I had multiple similar `adding_logs` tests and use this function to reduce code duplication. We can move this now into the test case directly. Thanks for the catch!

AudreyFelicio commented:
@anubh-v 
Thank you for the comments. I have pushed the changes.
Can you help to check if everything is fine now?

anubh-v commented:
Thanks for the contribution @AudreyFelicio , good work :+1:
